src/main/java/org/jitsi/utils/logging2/LogContext.java

```diff
@@ -16,53 +16,155 @@
 
 package org.jitsi.utils.logging2;
 
+import com.google.common.collect.*;
+import org.jetbrains.annotations.*;
+import org.jitsi.utils.collections.*;
+
 import java.util.*;
+import java.util.concurrent.*;
 import java.util.stream.*;
 
+/**
+ * Maintains a map of key-value pairs (both Strings) which holds
+ * arbitrary context to use as a prefix for log messages.  Sub-contexts
+ * can be created and will inherit any context values from their parent
+ * context.
+ */
+// Supress warnings about access since this is a library and usages will
+// occur outside this repo
+@SuppressWarnings("WeakerAccess")
 public class LogContext
 {
     public static LogContext EMPTY = new LogContext(Collections.emptyMap());
     public static String CONTEXT_START_TOKEN = "[";
     public static String CONTEXT_END_TOKEN = "]";
 
-    protected final Map<String, String> context;
-    protected final String formattedContext;
+    /**
+     * All context belonging to 'ancestors' of  this
+     * LogContext
+     */
+    protected ImmutableMap<String, String> parentContext;
+    /**
+     * The context held by this specific LogContext.
+     */
+    protected ImmutableMap<String, String> context;
+
+    /**
+     * The formatted String representing the total context
+     * (the combination of the parent context and this
+     * context)
+     */
+    protected String formattedContext;
+
+    /**
+     * Child LogContext's of this LogContext (which will be notified
+     * anytime this context changes)
+     */
+    private final List<LogContext> childContexts = new CopyOnWriteArrayList<>();
 
     public LogContext(Map<String, String> context)
     {
-        this.context = context;
-        this.formattedContext = formatContext(context);
+        this(context, ImmutableMap.of());
     }
 
-    protected String formatContext(Map<String, String> context)
+    protected LogContext(Map<String, String> context, ImmutableMap<String, String> parentContext)
     {
-        if (context.isEmpty()) {
+        this.context = ImmutableMap.copyOf(context);
+        this.parentContext = parentContext;
+        updateFormattedContext();
+    }
+
+    protected void updateFormattedContext()
+    {
+        this.formattedContext = formatContext(combineMaps(parentContext, context));
```

Comment:
To make sure I understand the Java thread model, this assignment is atomic, so there aren't any multi-threading issues here?  (It's a data race whether other threads get the old or new context, but that's not a problem.)

Reply:
right